runtime/src/cost_tracker.rs

```diff
@@ -82,71 +120,37 @@ impl CostTracker {
             return;
         }
 
-        let (costliest_account, costliest_account_cost) = self.find_costliest_account();
-
         datapoint_info!(
             "cost_tracker_stats",
             ("bank_slot", bank_slot as i64, i64),
             ("block_cost", self.block_cost as i64, i64),
             ("transaction_count", self.transaction_count as i64, i64),
-            (
-                "number_of_accounts",
-                self.cost_by_writable_accounts.len() as i64,
-                i64
-            ),
-            ("costliest_account", costliest_account.to_string(), String),
-            ("costliest_account_cost", costliest_account_cost as i64, i64),
         );
     }
 
-    fn find_costliest_account(&self) -> (Pubkey, u64) {
-        let mut costliest_account = Pubkey::default();
-        let mut costliest_account_cost = 0;
-        for (key, cost) in self.cost_by_writable_accounts.iter() {
-            if *cost > costliest_account_cost {
-                costliest_account = *key;
-                costliest_account_cost = *cost;
-            }
-        }
-
-        (costliest_account, costliest_account_cost)
-    }
-
-    fn would_fit(&self, keys: &[Pubkey], cost: &u64) -> Result<(), CostTrackerError> {
+    fn would_fit(
+        &self,
+        read_keys: &[Pubkey],
+        write_keys: &[Pubkey],
+        cost: u64,
+    ) -> Result<(), CostTrackerError> {
         // check against the total package cost
         if self.block_cost + cost > self.block_cost_limit {
             return Err(CostTrackerError::WouldExceedBlockMaxLimit);
         }
 
-        // check if the transaction itself is more costly than the account_cost_limit
-        if *cost > self.account_cost_limit {
-            return Err(CostTrackerError::WouldExceedAccountMaxLimit);
-        }
-
         // check each account against account_cost_limit,
-        for account_key in keys.iter() {
-            match self.cost_by_writable_accounts.get(&account_key) {
-                Some(chained_cost) => {
-                    if chained_cost + cost > self.account_cost_limit {
-                        return Err(CostTrackerError::WouldExceedAccountMaxLimit);
-                    } else {
-                        continue;
-                    }
-                }
-                None => continue,
-            }
+        if self.chained_cost.has_conflicts(read_keys, write_keys)
+            && self.account_cost_limit > self.chained_cost.total_chained_cost + cost
```

Comment:
`<=`, right?  Might pay to do the cheaper comparison first, too.

And `saturating_add`, while we're here :wink:

Reply:
yea, changed